src/mail.sh

```diff
@@ -0,0 +1,289 @@
+
+  if [[ "${options_values['SETUP']}" == 1 ]]; then
+    get_configs
+    mail_setup ''
+    exit
+  fi
+
+  return 0
+}
+
+# This function deals with configuring the options used by `git send-email`
+# to send patches by email
+#
+# Return:
+# returns 0 if successful, exits with 1 otherwise
+function mail_setup()
+{
+  local flag="$1"
+
+  flag=${flag:-'SILENT'}
+
+  is_inside_work_tree
+  if [[ "$?" -gt 0 ]] && [[ "${options_values['SCOPE']}" != 'global' ]]; then
+    complain 'Not in a git repository, aborting setup!'
+    say 'To apply settings globally rerun with "--global" flag.'
+    exit 22 # EINVAL
+  fi
+
+  for option in "${!options_values[@]}"; do
+    if [[ "$option" =~ smtp|user\.(email|name) ]]; then
+      add_config "$flag" "$option"
+    fi
+  done
+
+  return 0
+}
+
+# This function validates the encryption. If the passed encryption is not valid
+# this will warn the user and clear the option.
+#
+# @value: The value being passed, encryption must be ssl or tls
+#
+# Return:
+# Returns 0 if valid; 22 if invalid
+function validate_encryption()
+{
+  local value="$1"
+
+  if [[ "$value" =~ ^(ssl|tls)$ ]]; then
+    options_values['sendemail.smtpencryption']="$value"
+    return 0
+  fi
+
+  warning "Invalid encryption '$value', you must choose between 'ssl' or 'tls'."
+  warning 'Skipping this setting. This defaults to plain smtp.'
+
+  return 22 # EINVAL
+}
+
+# This function validates the encryption. If the passed encryption is not valid
+# this will warn the user and clear the option.
+#
+# @option: The option to determine if it should be an email
+# @value:  The value being passed
+#
+# Return:
+# Returns 0 if valid; 22 if invalid
+function validate_email()
+{
+  local option="$1"
+  local value="$2"
+
+  if [[ "$option" =~ ^(email|smtpuser)$ ]]; then
+    if [[ ! "$value" =~ ^[A-Za-z0-9_\.-]+@[A-Za-z0-9_-]+(\.[A-Za-z0-9]+)+$ ]]; then
+      complain "Invalid $option: $value"
+      return 22 #EINVAL
+    fi
+  fi
+
+  return 0
+}
+
+# This function adds the supplied configuration option. It first checks if the
+# given option has been set, then prompts the user to replace, add, or ignore
+# the configuration
+#
+# @option:     The option being edited
+# @value:      The value to update the option to
+# @curr_scope: The scope being edited
+# @cmd_scope:  The scope being imposed on the commands
+# @set_option: The relevant index to access set_confs
+#
+# Return:
+# Returns 0 if successful; non-zero otherwise
+function add_config()
+{
+  local flag="$1"
+  local option="$2"
+  local value="${options_values[$option]}"
+  local curr_scope=${options_values['SCOPE']}
+  local cmd_scope=${options_values['CMD_SCOPE']}
+  local set_option="${curr_scope}_$option"
+  local cmd
+
+  flag=${flag:-'SILENT'}
+
+  if [[ "${options_values['FORCE']}" == 0 ]]; then
+    if [[ -n "${set_confs[$set_option]}" ]]; then
+      warning "The configuration $option is already set with the following value(s):"
+      for scp in {'global','local'}; do
```

Comment:
scp? What is the meaning of this? Consider replacing it with a better name and declare it as a local variable.